gh workflow view

vilmibm opened the pull request:
Part of #2889 
Closes #3281 

This PR implements `gh workflow view`, a command for getting to a workflow yaml file.

With no arguments, interactively select a workflow:

![image](https://user-images.githubusercontent.com/98482/112215783-00f52b00-8bde-11eb-82f0-8e8643dee2fa.png)


Workflows are printed with syntax highlighting:

![image](https://user-images.githubusercontent.com/98482/111814711-ccfdcb00-88a8-11eb-9db5-6fd9997aa1b8.png)

Workflows can also be looked up by name, with interactive de-duping:

![image](https://user-images.githubusercontent.com/98482/112215835-0fdbdd80-8bde-11eb-9e5b-98929dec3adc.png)

Non-interactively, the just a workflow's yaml is printed raw. Non-unique name look ups result in an error:

![image](https://user-images.githubusercontent.com/98482/111814513-8c9e4d00-88a8-11eb-8cd0-98a338198422.png)

As with other commands, a raw ID can be used to look up a workflow. You can also pass a workflow's file name (ie `blank.yml`) as a workflow selector.

`--web` is supported, opening a page like https://github.com/vilmibm/testing/actions/workflows/blank.yml

## NB: spam workflows

It's possible that a workflow listed as belonging to a parent repo is:

- in an unmerged PR
- in a fork
- on a branch with the same name as the default branch of the parent repository

The actions API doesn't make it possible to distinguish these workflows from "real" workflows in a given repository, which means that an attempt to view them results in a 404. I both could not think of a way to fix the 404 nor think of a justification for fixing it, as in the experience of `cli/cli` all the workflows like this are spam. I think this issue only affects large, public repositories with spam and the path to mitigation for them are to disable those spam workflows (as I started doing in cli/cli).

vilmibm committed:
gh workflow view

samcoe commented on pkg/cmd/workflow/view/http.go:
```diff
@@ -0,0 +1,76 @@
+}
+
+func getWorkflowsByName(client *api.Client, repo ghrepo.Interface, workflowName string) ([]shared.Workflow, error) {
+	workflows, err := shared.GetWorkflows(client, repo, 100)
```
Are we worried about the 100 limit here? Should we be retrieving all workflows?

vilmibm replied:
yeah, we should probably be doing that.

samcoe commented on pkg/cmd/workflow/view/view.go:
```diff
@@ -0,0 +1,253 @@
+		}
+	}
+
+	if workflow == nil {
```
Would this be equivalent to `!opts.Prompt`? I think that stating the condition in that manner more clearly shows that this is the non-interactive path.

samcoe commented on pkg/cmd/workflow/view/view.go:
```diff
@@ -0,0 +1,253 @@
+}
+
+func promptWorkflows(client *api.Client, repo ghrepo.Interface) (*shared.Workflow, error) {
+	workflows, err := shared.GetWorkflows(client, repo, 10)
```
What was the reasoning behind limiting this to 10 workflows?

vilmibm replied:
totally arbitrary limit, i can bump it.

samcoe commented on pkg/cmd/workflow/view/view.go:
```diff
@@ -0,0 +1,253 @@
+		return &workflows[0], nil
+	}
+
+	if !io.CanPrompt() {
```
This is just opinion but it feels like this function could be split into two. One for retrieving the workflow(s) and then a second one that can get called from `runView` if more than one workflow is retrieved.

samcoe commented on pkg/markdown/markdown.go:
```diff
@@ -9,6 +9,24 @@ import (
 
 type RenderOpts []glamour.TermRendererOption
 
+func WithoutIndentation() glamour.TermRendererOption {
```
I like how you added these in a composable way!

Commit:
cleanup

- refactor resolveWorkflow to be more functions
- DRY the workflow prompting out
- use path.Base(workflow.Path) in place of path + ID in prompts
- preemptively move stuff into shared since the next workflow PR will
  make use of it

Commit:
support fetching all workflows

vilmibm commented:
NB: i've said that i was only able to fetch workflows for the default branch of a repo; turns out i can see other branches' workflows after all. i had a thinko in my code that was truncating my API results and accidentally led to me only seeing `trunk` workflows.

Commit:
support viewing by workflow filename

vilmibm commented:
@samcoe This is ready for re-review. I took care of your feedback and made one additional change: I properly support looking up workflows by their filename (which, to the API, is synonymous with their ID).

I also tweaked how workflows show up in the prompt like we discussed.

vilmibm commented:
I'll clean this up once #3244 is merged.

samcoe commented on pkg/cmd/workflow/shared/shared.go:
```diff
@@ -0,0 +1,157 @@
+		return nil, errors.New("empty workflow selector")
+	}
+
+	workflow, err := getWorkflowByID(client, repo, workflowSelector)
```
Is there a way to determine if the workflow selector is an ID or a name, outside of making the API request and checking for an error?

samcoe commented on pkg/cmd/workflow/view/view.go:
```diff
@@ -0,0 +1,195 @@
+		if opts.IO.IsStdoutTTY() {
+			fmt.Fprintf(opts.IO.Out, "Opening %s in your browser.\n", utils.DisplayURL(workflowURL))
+		}
+		return utils.OpenInBrowser(workflowURL)
```
Just a heads up since this will likely conflict with https://github.com/cli/cli/pull/3279, who ever merges second will have to resolve the conflict.

samcoe commented on pkg/cmd/workflow/shared/shared.go:
```diff
@@ -0,0 +1,157 @@
+	return workflows, nil
+}
+
+func SelectWorkflow(workflows []Workflow, promptMsg string, states []WorkflowState) (*Workflow, error) {
```
From what I can tell the calls to this function only use `active` as the `states` argument, do you foresee this being used with other states in the future?

samcoe reviewed (approved):
Left a couple questions but nothing blocking. This looks good 👍

vilmibm commented:
Unfortunately this is going to have to change considerably; I'm closing the PR for now.

Namely, this doesn't address how ephemeral Actions sees workflow files as; there is no consistent yaml file we can point to for a given workflow. A workflow is more of a temporal artifact of a run.

I'm going to be redoing this command as a summary of runs for a given workflow; actually fetching yaml will have to be rethought.
